ext/openssl/ossl_pkey.c

```diff
@@ -457,6 +457,28 @@ ossl_pkey_check_public_key(const EVP_PKEY *pkey)
     ossl_raise(ePKeyError, "public key missing");
 }
 
+static VALUE
+ossl_pkey_match(VALUE self, VALUE other)
+{
+  int ret;
+  EVP_PKEY *selfPKey;
+  EVP_PKEY *otherPKey;
+
+  GetPKey(self, selfPKey);
+  GetPKey(other, otherPKey);
+
+  ret = EVP_PKEY_cmp(selfPKey, otherPKey);
+
+  if (ret == -1)
+    ossl_raise(rb_eTypeError, "cannot match different PKey types");
```

Comment:
`ret == -1` doesn't necessarily mean the type mismatch but can generic error inside their `param/pub_cmp` implementations. We should check the type mismatch with `EVP_PKEY_id(pkey1) != EVP_PKEY_id(pkey2)` beforehand.

Reply:
Check my comment here https://github.com/ruby/openssl/pull/383#issuecomment-655660539. At the top `EVP_PKEY_cmp` does just that.